core/container.go

```diff
@@ -286,26 +286,64 @@ func (container *Container) Mounts(ctx context.Context) ([]string, error) {
 }
 
 func (container *Container) Directory(ctx context.Context, gw bkgw.Client, dirPath string) (*Directory, error) {
+	dir, err := locatePath(ctx, container, dirPath, gw, NewDirectory)
+	if err != nil {
+		return nil, err
+	}
+
+	// check that the directory actually exists so the user gets an error earlier
+	// rather than when the dir is used
+	_, err = dir.Stat(ctx, gw, ".")
+	if err != nil {
+		return nil, err
+	}
+
+	return dir, nil
+}
+
+func (container *Container) File(ctx context.Context, gw bkgw.Client, dirPath string) (*File, error) {
+	file, err := locatePath(ctx, container, dirPath, gw, NewFile)
+	if err != nil {
+		return nil, err
+	}
+
+	// check that the file actually exists so the user gets an error earlier
+	// rather than when the file is used
+	_, err = file.Stat(ctx, gw)
```

Comment:
Would this error out if the path was actually pointing to a dir?

Also it's worth double-checking what happens if the path is a symlink. I'm not even sure what the *correct* behavior should be then but we at least should know what happens currently.

Reply:
Good idea, it won't error as-is but it's an easy check against the `Stat` return value. I'll add the opposite too (requested dir, found file).

> Also it's worth double-checking what happens if the path is a symlink. I'm not even sure what the correct behavior should be then but we at least should know what happens currently.

So I wanted to try supporting this by just looping until the `Stat` result points to a value with an empty `Linkname`, but it looks like it just blows up instead:

```
input:7: container.from.withMountedDirectory.exec.exec.file stat file: lstat /tmp/buildkit-mount726192478/mnt/dir/symlink-1: no such file or directory
```

```go
func TestContainerFileSymlinks(t *testing.T) {
	t.Parallel()

	id := dirWithFileID(t, "final-file", "some-content")

	var linkRes struct {
		Container struct {
			From struct {
				WithMountedDirectory struct {
					Exec struct {
						Exec struct {
							File struct {
								Contents string
							}
						}
					}
				}
			}
		}
	}

	err := testutil.Query(
		`query Test($id: DirectoryID!) {
			container {
				from(address: "alpine:3.16.2") {
					withMountedDirectory(path: "/mnt/dir", source: $id) {
						exec(args: ["ln", "-s", "/mnt/dir/final-file", "/mnt/dir/symlink-1"]) {
							exec(args: ["ln", "-s", "/mnt/dir/symlink-1", "/mnt/dir/symlink-2"]) {
								file(path: "/mnt/dir/symlink-2") {
									contents
								}
							}
						}
					}
				}
			}
		}`, &linkRes, &testutil.QueryOptions{Variables: map[string]any{
			"id": id,
		}})
	require.NoError(t, err)
	require.Equal(t, "some-content", linkRes.Container.From.WithMountedDirectory.Exec.Exec.File.Contents)
}
```

I also tried `file(path: "/mnt/dir/symlink-1")` and got the same failure, but against `final-file`. Not sure why this is happening, it must be a dark corner of LLB.

Reply:
Oh huh I think this error message actually makes sense: `lstat /tmp/buildkit-mount726192478/mnt/dir/symlink-1`

What's happening is that you have a directory at `/mnt` with these contents:
```
/symlink1 -> /mnt/dir/final-file
/symlink2 -> /mnt/dir/symlink-1
```

But then when you are trying to stat the symlink it's on the Directory (it's not mounted anymore), so `/mnt/dir` doesn't exist anymore... So LLB's behavior is *technically* correct here I think, it's just really tricky code we have to write to handle maintaining the right paths to stat.

This isn't a blocker, but I do think we should fix this. It's not a common case but probably not obscure enough to ignore for all that long (IMO).

Reply:
Gotcha. The behavior I hoped for was that `Stat` wouldn't even attempt to follow symlinks, and instead I could just see where it's trying to go via the `Linkname` field in the return value and then follow it by going back out and through the regular mount lookup all over again, until finally the link is found.

What's confusing is it says `lstat` in the error which seems like the right thing to do, but it's erroring nonetheless!

Reply:
Oh yeah it should definitely be doing Lstat ([code here](https://github.com/sipsma/buildkit/blob/61307b8f2e73d681721b51856c10cb8606e5ea70/vendor/github.com/tonistiigi/fsutil/stat.go#L59-L59)). I suspect the problem is that we were trying to then stat the full `/mnt/dir/symlink-1` on the Directory at `/mnt/dir`, but the problem is that `/mnt/dir/symlink-1` only makes sense in the Container. In the Directory by itself the equivalent path is just `/symlink-1`

Reply:
Opened https://github.com/dagger/dagger/issues/3324 - merging!